src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -121,42 +170,99 @@ public RequestExecutionQueue(ILspSolutionProvider solutionProvider)
+                            ranToCompletion = await work.CallbackAsync(context, cancellationToken).ConfigureAwait(false);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                            // Since the cancellationToken passed to the callback is a linked source it could be cancelled
+                            // without the queue token being cancelled, but ranToCompletion will be false so no need to
+                            // do anything special here.
+                        }
 
-                    // If the handling of the request failed, the exception will bubble back up to the caller, but we
-                    // still need to react to it here by throwing away solution updates
-                    if (ranToCompletion)
+                        // If the handling of the request failed, the exception will bubble back up to the caller, but we
+                        // still need to react to it here by throwing away solution updates
+                        if (ranToCompletion)
+                        {
+                            lastMutatedSolution = mutatedSolution ?? lastMutatedSolution;
+                        }
+                        else
+                        {
+                            OnRequestServerShutdown($"An error occured processing a mutating request and the solution is in an invalid state. Check LSP client logs for any error information.");
+                            break;
+                        }
+                    }
+                    else
                     {
-                        lastMutatedSolution = mutatedSolution ?? lastMutatedSolution;
+                        // Non mutating are fire-and-forget because they are by definition readonly. Any errors
+                        // will be sent back to the client but we can still capture errors in queue processing
+                        // via NFW, though these errors don't put us into a bad state as far as the rest of the queue goes.
+                        _ = work.CallbackAsync(context, cancellationToken).ReportNonFatalErrorAsync();
                     }
                 }
-                else
-                {
-                    // Non mutating request get given the current solution state, but are otherwise fire-and-forget
-                    _ = work.CallbackAsync(context, cancellationToken);
-                }
+            }
+            catch (Exception e) when (FatalError.ReportWithoutCrash(e))
+            {
+                OnRequestServerShutdown($"Error occured processing queue: {e.Message}.");
+            }
+        }
+
+        private void OnRequestServerShutdown(string message)
+        {
+            RequestServerShutdown?.Invoke(this, new RequestShutdownEventArgs(message));
+
+            if (_cancelSource != null)
+            {
+                _cancelSource.Cancel();
+                DrainQueue();
+            }
+        }
```

Comment:
all these shoudl have some rich comments explaining what is going on and what state this leaves thigns in.

Reply:
Depending on how "rich" you are talking about, I might have done this :)